### ISSUE TITLE
CSLA v5 and unit testing BusinessRules

### ISSUE THREAD

**Issue by GillesBer** [MarimerLLC/cslaforum#790]
**Question**
I'm playing with the prerelease bits of CSLA v5.
I'm not able to compile my business rules test. I do something similar to:
https://github.com/MarimerLLC/csla/blob/master/Samples/NET/cs/RuleTutorial/RuleTutorial.Testing.Common/BusinessRuleTest.cs ln. 57
`RuleContext.Rule.Execute(RuleContext);`

RuleContext.Rule is now a IBusinessRuleBase (was a IBusinessRule before).
It no longer implements Execute .
How should we test our rules, now?

I suspect that IBusinessRuleBase was introduced by https://github.com/MarimerLLC/csla/issues/409
Should we check if 'IsAsync' then explicitely cast to IBusinessRule or IBusinessRuleAsync
Regards
**Version and Platform**
CSLA version: 5.0.0.0 prerelease   NET4.6
OS: Windows,
Platform: WinForms, WPF

**Comment by rockfordlhotka** [comment]
Good point @GillesBer - I'm not sure whether some helper code is a good idea?

The way to invoke a rule is now like this:

```c#
          if (rule is IBusinessRule syncRule)
            syncRule.Execute(context);
          else if (rule is IBusinessRuleAsync asyncRule)
            asyncRule.ExecuteAsync(context).ContinueWith((t)=> { context.Complete(); });
```

At least that's the [code in the BusinessRules class](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Rules/BusinessRules.cs#L771).

I'm not sure it is ideal for every unit test to have that code though? So maybe there should be a static helper method that `BusinessRules` and unit tests can invoke to execute a rule?

**Comment by GillesBer** [comment]
Thanks, I got it, and my test now compiles.
I have now to check what gets wrong :).
Thanks.

**Comment by rockfordlhotka** [comment]
What do you think about the helper method idea? Valuable or not?

**Comment by GillesBer** [comment]
In our case, we use a base class for our unit tests.
We referered to the RuleTutorial to know how to built it.
Because there is much more to know/understand that just this change.

So, to start from scratch, I would rather prioritize to upgrade the RuleTutorial, as someone will anyway need to be aware of what this change implies.

The value of the extension method is that this change will block additional members of the community (and other will raise issue like mine), so this is valuable to turn this change as not blocking.
